### DISCUSSION TITLE
Possible schema ideas

### DISCUSSION THREAD

**Discussion by the-mentalist** [TrakGene/react-form-builder#7] [Ideas]
Hi @Shubhrajyoti-Dey-FrosTiK 

### Initial idea
I had this schema as part of the POC, but recently after discussing it with Tim turns out this won't fit our requirements.
I am currently doing a bit of research around this and would love to hear your ideas.

This is what I thought
An object, that will contain basic information about the form like `title`, `description` or `width`. It contains a property called `elements` that refers to an array of form input elements. Each element has a unique `id` (something like UUID) that is used to store the information corresponding to that `id`
For instance
```
{
  title: 'Title of the page',
  description: 'description of the page',
  width: 4,
  elements: [{
    id: 'u1',
    type: TYPES.textField,
    label: 'Name of Clinical Indication',
  }, {
    id: 'u2',
    type: TYPES.radio,
    label: 'Testing for Familial Variant',
    options: [{value: 'y', label: 'Yes'}, {value: 'n', label: 'No'}],
    row: true
  }, {
    id: 'u3',
    type: TYPES.textField,
    label: 'Name of person whose sample is to be tested',
  }, {
    id: 'u4',
    type: TYPES.date,
    label: 'Date of birth',
    disableFuture: true
  }, {
    id: 'u5',
    type: TYPES.textField,
    label: 'NHS number',
  }, {
    id: 'u6',
    type: TYPES.textField,
    label: 'Hospital Number',
  }, {
    id: 'u7',
    type: TYPES.staticList,
    label: 'Possible results:',
    options: [
      'A change may be found that is the whole or at least partial cause of the condition. It may have other health implications for me and / or other family members',
      'No differences may be found i.e.testing may not find the cause of the condition.',
      'A change may be found that cannot be interpreted with current scientific knowledge (further family studies or research may be helpful).',
      'In  rare  situations  a  change  may  be  found  that  is  not  thought  to  be  the  cause  of  the  condition  being investigated, but could have other implications for me and / or other family members.'
    ]
  }]
}
```

will generate something like this
![image](https://user-images.githubusercontent.com/3134785/173017854-1f6a1a8e-4b43-49dd-be37-22389705edce.png)

BUT

### Problems with this approach
- Cannot prefill information. Since everything is save based on the `id`, its hard to determine the meaning of the field, for instance in the above screenshot we want Name and DOB fields to be prefilled.
- Cannot combine forms, for instance, the ability to create a demographic element that contains multiple fields related to patients' demography and ability to add this to another form

@tjackson-trakgene feel free to add if I have missed something

**Comment by the-mentalist** [comment]
How to implement a condition or dependency tree within elements?

**Reply by Shubhrajyoti-Dey-FrosTiK** [reply]
I have a suggestion regarding this 

For this approach please keep aside the `condition handling` (`if`, `else`, `equality` etc). In this approach we will be looking into how we can render `conditionally`. 

So lets denote this `condition handling`  by `CONDITION`

***Intuition***
User needs to make forms which are personalised for each user. So  lets consider an example

```.js
GENDER : MALE -> Ask the Date of Birth
GENDER : FEMALE -> Ask Mothers Name
``` 
and there are some questions which has to asked to both of them 

```.js
NAME: ______
COUNTRY: _______
```

So what this intuition says is to group these form elements.

Grouping them: 

```.js
// Group 1
NAME: _______
COUNTRY: _______
GENDER: _________

// Group 2
DATE_OF_BIRTH:  ______

// Group 3
MOTHERS_NAME: _____
```

Now clearly we can see `Group 2` and `Group 3` is conditionally dependent on `Group 1`
Or in other words `Group 1` diverges into two path `Group 2` and `Group 3`

So`Group 1` structure will look like this:

```.js
{
  description: "FORM_DESCRIPTION"
  formElements: [{
      id: 'u1',
      type: TYPES.textField,
      label: 'Name',
    }, {
      id: 'u1',
      type: TYPES.textField,
      label: 'Country',
    }]
  groupsConnectedTo: [
     {id: "GROUP_2_ID", renderType: "conditional", condition : CONDITION},
     {id: "GROUP_3_ID", renderType: "conditional", condition : CONDITION}
  ],
  id: "GROUP_1_ID"
  previousConnections: [null]
  renderType: "default"
  title: "FORM_TITLE"
}
```
`previousConnections`  contains all the groups which have `attached` this group in their flow. 

`renderType` can be `DEFAULT` or `CONDITIONAL` 
if `CONDITIONAL`, a key `condition` has to be passed where the conditions will be specified

If `groupsConnectedTo` is `[]` then it means that it is the end of the `form flow`

Now lets consider this above code as `GROUP_1` schema

So the final structure will look like this : 

```.js
{
   formTitle: FORM_TITLE,
   startingGroupId: GROUP_1.id,
   description: FORM_DESCRIPTION
   schema: {
       GROUP_1.id: GROUP_1,
       GROUP_2.id: GROUP_2,
       .....
    }
}
```
`startingGroupId` denotes the first `Group` to be rendered.

So this is basically a **`A directed acyclic Graph (TREE) data structure`** which we are implementing.

***Why use this type of a structure?***
1. Easier to store data.
2. Accessing a group takes `O(1)` time.
3. Adding FormElement takes `O(1)` time.
4. Deleting a FormElement takes `O(n)` time where `n` is the number of formElements in the group.
5. Deleting a `Group` takes `O(n)` time where `n` is the  number of `Groups` which have attached this concerned group in the `form flow`
6. User gets a more visual interpretation how the flow is working.

***Not Implemented***
Need to look into `condition handling` a bit.

**Comment by the-mentalist** [comment]
Heya @Shubhrajyoti-Dey-FrosTiK 
A quick note and maybe you can add that to your schema structure:

To handle the scenario:

> user info is **prefilled** and can be reused on other forms

for instance, for email or dob fields, I have introduced a new field "_user_" to the element structure

For example
```
{
    ...
    "user": "email"
    ...
}
```

This allows me to extract value from the response and update the user table and prefill the custom form.


Right now, these values for _user_ property will depend on the [user schema](https://github.com/TrakGene/trakgenecloud/blob/main/backend/models/form-users.js#L32), possible values atm can be email & dob

cc: @tjackson-trakgene
